### ISSUE TITLE
large partition  std::bad_alloc

### ISSUE THREAD

**Issue by asias** [scylladb/scylla#1545]
export SCYLLA_EXT_OPTS="--smp 2 --memory 1G --ring-delay-ms 5000"
update_cluster_layout_tests.py:TestUpdateClusterLayout.add_node_with_large_partition1_test

When increasing the nr_rows from 100 to 1000 in the test, the test fails

```
--- a/update_cluster_layout_tests.py
+++ b/update_cluster_layout_tests.py
@@ -1195,7 +1195,7 @@ class TestUpdateClusterLayout(Tester):
         self.create_ks(session, 'ks', 2)
         self.create_cf(session, 'cf', read_repair=0.0, columns={'c1': 'text', 'c2': 'text'})

-        nr_rows = 100
+        nr_rows = 1000
         c1 = 'a' * 1024 * 100 # 100KB
         c2 = 'b' * 1024 * 300 # 300KB
         c1s = [c1] * nr_rows
```

On node2:

```
  0 ERROR 2016-08-04 14:44:25,216 [shard 1] storage_proxy - Exception when communicating with 127.0.0.2: std::bad_alloc
  1 INFO  2016-08-04 14:44:26,008 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
  2 INFO  2016-08-04 14:44:27,008 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
  3 INFO  2016-08-04 14:44:28,008 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
  4 INFO  2016-08-04 14:44:29,008 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
  5 INFO  2016-08-04 14:44:30,008 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
  6 INFO  2016-08-04 14:44:31,008 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
  7 INFO  2016-08-04 14:44:32,009 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
  8 INFO  2016-08-04 14:44:33,009 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
  9 INFO  2016-08-04 14:44:34,009 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 10 INFO  2016-08-04 14:44:35,009 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 11 ERROR 2016-08-04 14:44:35,215 [shard 1] storage_proxy - Exception when communicating with 127.0.0.2: std::bad_alloc
 12 INFO  2016-08-04 14:44:36,009 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 13 INFO  2016-08-04 14:44:37,009 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 14 INFO  2016-08-04 14:44:38,010 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 15 INFO  2016-08-04 14:44:39,010 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 16 INFO  2016-08-04 14:44:40,010 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 17 INFO  2016-08-04 14:44:41,010 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 18 INFO  2016-08-04 14:44:42,010 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 19 INFO  2016-08-04 14:44:43,010 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
 20 INFO  2016-08-04 14:44:44,011 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
```

scylla commit: 0c1bf6c scylla-gdb.py: Fix lookup of global symbols
dtest commit: 87876da Merge pull request #375 from asias/add_node_with_large_partition_test

**Comment by asias** [comment]
update: dropping the extra "export SCYLLA_EXT_OPTS="--smp 2 --memory 1G --ring-delay-ms 5000" env, it still fails with std::bad_alloc.

**Comment by tgrabiec** [comment]
Can you try to create a coredump when bad_alloc happens (gdb: catch throw std::bad_alloc)?

**Comment by asias** [comment]
On Aug 4, 2016 3:29 PM, "Tomasz Grabiec" <notifications@github.com> wrote:
>
> Can you try to create a coredump when bad_alloc happens (gdb: catch throw
std::bad_alloc)?

It is running inside dtest. Not sure if we can.

**Comment by tgrabiec** [comment]
You could start the cluster manually using ccm, then attach with gdb to it, insert breakpoints, then run the dtest. Note that GDB slows down Syclla a lot if you run on many CPUs. Maybe with -c 2 it's not tragic.

**Comment by tgrabiec** [comment]
I'm going to send a patch which adds `--abort-on-bad-alloc` which should help with that.

**Comment by asias** [comment]
During the test, scylla is started and stopped more than once. A better option is to insert a sleep before the operation which will cause the bad_alloc, using this time to attach gdb to scylla. I will try your --abort-on-bad-alloc patch.

**Comment by pdziepak** [comment]
It may be worth noting that this dtest creates partitions with a single clustering row containing two blobs one 100kB, the other 300kB. This is, obviously, the case in which streamed mutations bring no improvement.

The result is that we end up `_mutation_send_limiter` (=256) blobs of frozen unfragmented mutations each over 400kB in size. That's certainly going to put some pressure on memory allocator.

We probably should limit number of concurrent mutation send using frozen mutation size and not just number of mutations.

**Comment by asias** [comment]
@pdziepak I checked with the following patch. The mutation is indeed fragmented 
```
--- a/streaming/stream_transfer_task.cc
+++ b/streaming/stream_transfer_task.cc
@@ -119,8 +119,9 @@ future<> send_mutations(auto si) {
                     if (!service::get_local_storage_service().cluster_supports_large_partitions()) {
                         fragment_size = std::numeric_limits<size_t>::max();
                     }
-                    return fragment_and_freeze(std::move(*smopt), [si] (auto fm, bool fragmented) {
+                    return fragment_and_freeze(std::move(*smopt), [si, fragment_size] (auto fm, bool fragmented) {
                         si->mutations_nr++;
+                        sslog.info("size={}, fragmented={}, fragment_size={}", fm.representation().size(), fragmented, fragment_size);
                         return do_send_mutations(si, std::move(fm), fragmented);
                     }, fragment_size).then([] { return stop_iteration::no; });
                 } else {
```

Log shows mutation is fragmented:

```
INFO  2016-08-05 10:00:16,353 [shard 1] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,354 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,355 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,356 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,356 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,358 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,359 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,359 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,360 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,361 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,361 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,362 [shard 1] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,364 [shard 1] stream_session - size=409830, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,364 [shard 1] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,365 [shard 1] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,366 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,366 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,367 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,367 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] Session with 127.0.0.2 is complete, state=COMPLETE
INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] bytes_sent = 409831890, bytes_received = 0
INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] All sessions completed, peers={127.0.0.2}
```


Also, yhe bad_alloc does not happen during the streaming but after streaming is done (self.check_rows_on_node(node2, nr_rows)): 
```
RUN TEST: update_cluster_layout_tests.py:TestUpdateClusterLayout.add_node_with_large_partition1_test
add_node_with_large_partition1_test (update_cluster_layout_tests.TestUpdateClusterLayout) ... cluster ccm directory: /home/asias/.dtest/dtest-W20s9i
Node 1 started
Insert data
Node 2 started
Check rows on node2
Retrying read after timeout. Attempt #0
Retrying read after timeout. Attempt #1
Retrying read after timeout. Attempt #2
Retrying read after timeout. Attempt #3
Retrying read after timeout. Attempt #4
ERROR
ERROR

```
Log on node2:
```
INFO  2016-08-05 10:00:16,476 [shard 0] compaction - Compacting [/home/asias/.dtest/dtest-W20s9i/test/node2/data/ks/cf-524275d05ab011e6b6c1000000000001/ks-cf-ka-2-Data.db:level=0, /home/asias/.dtest/dtest-W20s9i/test/node
INFO  2016-08-05 10:00:16,605 [shard 1] compaction - Compacting [/home/asias/.dtest/dtest-W20s9i/test/node2/data/ks/cf-524275d05ab011e6b6c1000000000001/ks-cf-ka-3-Data.db:level=0, /home/asias/.dtest/dtest-W20s9i/test/node
INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] Session with 127.0.0.1 is complete, state=COMPLETE
INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] bytes_sent = 0, bytes_received = 409831890
INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] All sessions completed, peers={127.0.0.1}
INFO  2016-08-05 10:00:16,606 [shard 0] storage_service - Bootstrap completed! for the tokens {4635781158196633014, -5516070834301145815, -156176976896236884, -1171508881780488170, -1918184326712110280, -25216943561102467
INFO  2016-08-05 10:00:16,914 [shard 0] storage_service - Node 127.0.0.2 state jump to normal
INFO  2016-08-05 10:00:17,053 [shard 0] storage_service - NORMAL: node is now in normal status
INFO  2016-08-05 10:00:17,054 [shard 0] gossip - Checking if need_features {RANGE_TOMBSTONES} in features {LARGE_PARTITIONS, RANGE_TOMBSTONES}
INFO  2016-08-05 10:00:17,054 [shard 0] gossip - Checking if need_features {LARGE_PARTITIONS} in features {LARGE_PARTITIONS, RANGE_TOMBSTONES}
INFO  2016-08-05 10:00:17,054 [shard 1] gossip - Checking if need_features {RANGE_TOMBSTONES} in features {LARGE_PARTITIONS, RANGE_TOMBSTONES}
INFO  2016-08-05 10:00:17,054 [shard 1] gossip - Checking if need_features {LARGE_PARTITIONS} in features {LARGE_PARTITIONS, RANGE_TOMBSTONES}
INFO  2016-08-05 10:00:17,054 [shard 0] gossip - Waiting for gossip to settle before accepting client requests...
INFO  2016-08-05 10:00:19,543 [shard 1] compaction - Compacted 7 sstables to [/home/asias/.dtest/dtest-W20s9i/test/node2/data/ks/cf-524275d05ab011e6b6c1000000000001/ks-cf-ka-17-Data.db:level=0, ]. 195826039 bytes to 19582
INFO  2016-08-05 10:00:19,620 [shard 0] compaction - Compacted 8 sstables to [/home/asias/.dtest/dtest-W20s9i/test/node2/data/ks/cf-524275d05ab011e6b6c1000000000001/ks-cf-ka-18-Data.db:level=0, ]. 213851851 bytes to 21385
INFO  2016-08-05 10:00:25,054 [shard 0] gossip - No gossip backlog; proceeding
INFO  2016-08-05 10:00:25,054 [shard 0] storage_service - Starting listening for CQL clients on 127.0.0.2:9042...
INFO  2016-08-05 10:00:25,054 [shard 0] storage_service - Thrift server listening on 127.0.0.2:9160 ...
INFO  2016-08-05 10:00:29,386 [shard 0] gossip - InetAddress 127.0.0.1 is now DOWN
INFO  2016-08-05 10:00:31,488 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:32,488 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
ERROR 2016-08-05 10:00:33,196 [shard 1] storage_proxy - Exception when communicating with 127.0.0.2: std::bad_alloc
INFO  2016-08-05 10:00:33,488 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:34,489 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:35,489 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:36,489 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:37,489 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:38,489 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:39,489 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:40,489 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:41,490 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:42,490 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
ERROR 2016-08-05 10:00:42,820 [shard 1] storage_proxy - Exception when communicating with 127.0.0.2: std::bad_alloc
INFO  2016-08-05 10:00:43,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:44,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:45,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:46,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:47,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:48,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:49,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:50,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:51,491 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:52,492 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
ERROR 2016-08-05 10:00:52,820 [shard 1] storage_proxy - Exception when communicating with 127.0.0.2: std::bad_alloc
INFO  2016-08-05 10:00:53,492 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:54,492 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:55,492 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
INFO  2016-08-05 10:00:56,492 [shard 0] rpc - client 127.0.0.1: fail to connect: Connection refused
```

**Comment by asias** [comment]
backtrace on node2 with --abort-on-bad-alloc

```
(gdb) bt
#0  0x00007f7e21d1d6f5 in raise () from /lib64/libc.so.6
#1  0x00007f7e21d1f2fa in abort () from /lib64/libc.so.6
#2  0x0000000000502864 in memory::on_allocation_failure (size=<optimized out>, size@entry=409833) at core/memory.cc:89
#3  0x00000000005028e0 in memory::allocate (size=size@entry=409833) at core/memory.cc:975
#4  0x0000000000502a53 in malloc (n=n@entry=409833) at core/memory.cc:1116
#5  0x0000000000c2b401 in basic_sstring<signed char, unsigned int, 31u>::basic_sstring (x=..., this=0x601003aad2a0) at /home/asias/src/cloudius-systems/scylla/seastar/core/sstring.hh:174
#6  frozen_mutation::frozen_mutation (m=..., this=0x601003aad2a0) at ./frozen_mutation.hh:53
#7  partition::partition (this=0x601003aad298) at ./mutation_query.hh:33
#8  __gnu_cxx::new_allocator<partition>::construct<partition, partition const&> (this=0x601000617a18, __p=0x601003aad298) at /usr/include/c++/6.1.1/ext/new_allocator.h:120
#9  std::allocator_traits<std::allocator<partition> >::construct<partition, partition const&> (__a=..., __p=0x601003aad298) at /usr/include/c++/6.1.1/bits/alloc_traits.h:455
#10 std::vector<partition, std::allocator<partition> >::push_back (__x=..., this=0x601000617a18) at /usr/include/c++/6.1.1/bits/stl_vector.h:918
#11 service::mutation_result_merger::get()::{lambda()#2}::operator()() (__closure=__closure@entry=0x601000617a08) at service/storage_proxy.cc:3426
#12 0x0000000000c2b61f in futurize<future<std::experimental::fundamentals_v1::optional<reconcilable_result> > >::apply<service::mutation_result_merger::get()::{lambda()#2}&>(service::mutation_result_merger::get()::{lambda()#2}&) (func=...) at /home/asias/src/cloudius-systems/scylla/seastar/core/future.hh:1244
#13 repeat_until_value<service::mutation_result_merger::get()::{lambda()#2}>(std::result_of&&) (action=action@entry=<unknown type in /home/asias/src/cloudius-systems/scylla/build/release/scylla, CU 0xbcc9227, DIE 0xc1df95d>)
    at /home/asias/src/cloudius-systems/scylla/seastar/core/future-util.hh:266
#14 0x0000000000c2ccb6 in repeat_until_value_type_helper<std::result_of<service::mutation_result_merger::get()::{lambda()#2} ()>::type>::future_type repeat_until_value<service::mutation_result_merger::get()::{lambda()#2}>(std::result_of&&)::{lambda()#2}::operator()() (__closure=0x601000617a08) at /home/asias/src/cloudius-systems/scylla/seastar/core/future-util.hh:292
#15 lambda_task<repeat_until_value_type_helper<std::result_of<service::mutation_result_merger::get()::{lambda()#2} ()>::type>::future_type repeat_until_value<service::mutation_result_merger::get()::{lambda()#2}>(std::result_of&&)::{lambda()#2}>::run() (this=0x601000617a00) at /home/asias/src/cloudius-systems/scylla/seastar/core/task.hh:40
#16 0x0000000000484939 in reactor::run_tasks (this=this@entry=0x6010000c3000, tasks=...) at core/reactor.cc:1531
#17 0x00000000004b284b in reactor::run (this=0x6010000c3000) at core/reactor.cc:1951
#18 0x00000000004cf2de in smp::<lambda()>::operator()(void) const (__closure=0x600000159900) at core/reactor.cc:2825
#19 0x00000000004ffdae in std::function<void ()>::operator()() const (this=<optimized out>) at /usr/include/c++/6.1.1/functional:2136
#20 posix_thread::start_routine (arg=<optimized out>) at core/posix.cc:50
#21 0x00007f7e220b25ca in start_thread () from /lib64/libpthread.so.0
#22 0x00007f7e21debead in clone () from /lib64/libc.so.6
```

Patch to dtest to give time to attach gdb to node2.

```
--- a/update_cluster_layout_tests.py
+++ b/update_cluster_layout_tests.py
@@ -1207,6 +1207,10 @@ class TestUpdateClusterLayout(Tester):
         node2.start(wait_for_binary_proto=True)
         debug("Node 2 started")

+        debug("Sleep 60 seconds")
+        time.sleep(60)
+        debug("Wakeup")
+
```

**Comment by tgrabiec** [comment]
@asias Note that abort should trigger coredump creation so you don't need to attach GDB beforehand if you have coredumps configured properly

**Comment by pdziepak** [comment]
On 5 August 2016 at 03:08, Asias He <notifications@github.com> wrote:

> @pdziepak <https://github.com/pdziepak> I checked with the following
> patch. The mutation is indeed fragmented
>
> --- a/streaming/stream_transfer_task.cc
> +++ b/streaming/stream_transfer_task.cc
> @@ -119,8 +119,9 @@ future<> send_mutations(auto si) {
>                      if (!service::get_local_storage_service().cluster_supports_large_partitions()) {
>                          fragment_size = std::numeric_limits<size_t>::max();
>                      }
> -                    return fragment_and_freeze(std::move(*smopt), [si] (auto fm, bool fragmented) {
> +                    return fragment_and_freeze(std::move(*smopt), [si, fragment_size] (auto fm, bool fragmented) {
>                          si->mutations_nr++;
> +                        sslog.info("size={}, fragmented={}, fragment_size={}", fm.representation().size(), fragmented, fragment_size);
>                          return do_send_mutations(si, std::move(fm), fragmented);
>                      }, fragment_size).then([] { return stop_iteration::no; });
>                  } else {
>
> Log shows mutation is fragmented:
>
> INFO  2016-08-05 10:00:16,353 [shard 1] stream_session - size=409832, fragmented=1, fragment_size=131072
>
> Right, I though that `fragment_and_freeze()` doesn't set `fragmented` flag
if the next mutation fragment is end of stream, I must've misremembered.
Anyway, it still produces fragments ~400kB in size which puts memory
allocator under pressure.


> INFO  2016-08-05 10:00:16,354 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,355 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,356 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,356 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,358 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,359 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,359 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,360 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,361 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,361 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,362 [shard 1] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,364 [shard 1] stream_session - size=409830, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,364 [shard 1] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,365 [shard 1] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,366 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,366 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,367 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,367 [shard 0] stream_session - size=409832, fragmented=1, fragment_size=131072
> INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] Session with 127.0.0.2 is complete, state=COMPLETE
> INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] bytes_sent = 409831890, bytes_received = 0
> INFO  2016-08-05 10:00:16,606 [shard 0] stream_session - [Stream #599cefe0-5ab0-11e6-8b70-000000000000] All sessions completed, peers={127.0.0.2}
>
> Also, yhe bad_alloc does not happen during the streaming but after
> streaming is done (self.check_rows_on_node(node2, nr_rows)):
>
Right, that's because self.check_fows_on_node() is even worse for the
allocator than the streaming itself. It asks for `SELECT * FROM cf`.
Default python driver page size appears to be 5000. So the coordinator
needs to allocate 1000 continuous blobs 400kB each.

**Comment by avikivity** [comment]
We're allowed to use smaller pages than specified by the driver, we should exploit that.

**Comment by tgrabiec** [comment]
@pdziepak The exception is thrown on frozen_mutation construction. It currently uses bytes as storage, so we will need contiguous memory for the whole partition in the query results. Perhaps we should switch to bytes_ostream?

**Comment by pdziepak** [comment]
Switching `frozen_mutation` to `bytes_ostream` certainly would help and a patchset doing that is going to be sent soon.

Note, however, that the node in a cluster is given 1 GB of memory and there are two shards. So we have 512 MB of memory on coordinator which has to sent over 400 MB response to the client. Even if the buffer is fragmented that is going to result in allocation failure. The only solution in such cases is to return less rows than the coordinator was asked for if the paging is enabled or fail with `std::bad_alloc` if it's not.

**Comment by avikivity** [comment]
Limiting response sizes is good practice anyway.

If we limit pages to 1MB we can still do 1GB/s with 1ms round-trip times.  Since we do expect some concurrency, and we 1GB/s is really optimistic, I think it won't harm throughput too much.

**Comment by tzach** [comment]
@pdziepak @avikivity anything else to do for this issue?

**Comment by avikivity** [comment]
Lots

**Comment by Jan-M** [comment]
Not sure if this is related, but my node drops gazillion of log lines like this: (amounting to 1.3GB of logs the last hour)

```
Sep  5 14:08:35 ip-172-31-128-11 docker/0a2f2d6f525b[828]: WARN  2016-09-05 14:08:35,178 [shard 0] storage_proxy - Failed to apply mutation from 172.31.144.12#3: std::bad_alloc (std::bad_alloc)
Sep  5 14:08:35 ip-172-31-128-11 docker/0a2f2d6f525b[828]: WARN  2016-09-05 14:08:35,178 [shard 3] storage_proxy - Failed to apply mutation from 172.31.144.11#3: std::bad_alloc (std::bad_alloc)
Sep  5 14:08:35 ip-172-31-128-11 docker/0a2f2d6f525b[828]: WARN  2016-09-05 14:08:35,178 [shard 0] storage_proxy - Failed to apply mutation from 172.31.144.12#3: std::bad_alloc (std::bad_alloc)
Sep  5 14:08:35 ip-172-31-128-11 docker/0a2f2d6f525b[828]: WARN  2016-09-05 14:08:35,178 [shard 3] storage_proxy - Failed to apply mutation from 172.31.144.11#3: std::bad_alloc (std::bad_alloc)
Sep  5 14:08:35 ip-172-31-128-11 docker/0a2f2d6f525b[828]: WARN  2016-09-05 14:08:35,178 [shard 0] storage_proxy - Failed to apply mutation from 172.31.144.12#3: std::bad_alloc (std::bad_alloc)
Sep  5 14:08:35 ip-172-31-128-11 docker/0a2f2d6f525b[828]: message repeated 2 times: [ WARN  2016-09-05 14:08:35,178 [shard 0] storage_proxy - Failed to apply mutation from 172.31.144.12#3: std::bad_alloc (std::bad_alloc)]
```

On AWS EC2 with 12GB of 16GB assigned and 4 cores. 

Anything I need to do?

**Comment by avikivity** [comment]
Fixed via #1633.
